### PR TITLE
Book

### PR DESCRIPTION
**Pull request by frondeus**
Deployed to: https://epiphany-platform.github.io/CommonDataLayer/

### CODE REVIEW

**Comment by smores56** [comment]
Can you update the GitHub Action that automatically deploys this? You have to rename the path it checks from `book` to `docs`. https://github.com/epiphany-platform/CommonDataLayer/blob/develop/.github/workflows/deploy-mdbook.yml

File: docs/architecture/frontend.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Admin Web Panel
+
+This is the management portal for the CDL, useful for updating
+configuration, manipulating data, and viewing performance data.
```
Comment (Kononnable) [inline docs/architecture/frontend.md:4]:
```suggestion
configuration, manipulating data.
```
I think performance(metrics) are out of its scope for now.

File: docs/architecture/ingestion.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Ingestion Layer
+
+Services in this layer are responsible for accepting generic messages from external systems via `Kafka`, validating them and sorting to correct repository.
```
Comment (Kononnable) [inline docs/architecture/ingestion.md:3]:
```suggestion
Services in this layer are responsible for accepting generic messages from external systems via a message queue, validating them and and forwarding the message to correct repository.  
```

File: docs/architecture/query_router.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# Query Router
+
+### Technical Description
+
+The Query Router (`QR`), is responsible for finding specific query service. Since messages can be stored in any available repository, data router acts as a single entry point to multi-repo system. Query Router first queries SR, then basing on received config, finds out specific QS that, hopefully, should be able to respond to specific query. Logic of that process is based on repo_type and query-service address stored with schema itself.
```
Comment (Kononnable) [inline docs/architecture/query_router.md:5]:
```suggestion
The Query Router (`QR`), is responsible for forwarding requests to specific query services. In CDL messages can be stored in any available repository, data router acts as a single entry point to multi-repo system and query router allows that data to be fetched easily.
Query Router first queries SR, then basing on received config, finds out specific QS that, hopefully, should be able to respond to specific query. Logic of that process is based on repo_type and query-service address stored with schema itself.
```

File: docs/architecture/query_service.md [hunk 4]
```diff
@@ -1,12 +1,14 @@
...
+Two query-services are present: one for timeseries databases, and one for documents.
 
 ### Technical Description
 
 The query service (`QS` or for example for postgresql its `QSPG`), is responsible for querying data from specific repository. It offers two paths that can be accessed:
 First path depends on type of repo
 
 ### Communication
-Communication to query service is done through [gRPC][grpc] based on two [endpoints][endpoints] of querying for data by `SCHEMA_ID` or multiple `OBJECT_ID`s. Query service communicates with multiple databases such as postgresql, druid, sled. Query service also communicates with [schema registry][schema-registry]. 
+Communication to query service is done through [gRPC][grpc] based on two [endpoints][proto] of querying for data by `SCHEMA_ID` or multiple `OBJECT_ID`s. Query service communicates with multiple databases such as postgresql, druid, sled. Query service also communicates with [schema registry][schema-registry].
```
Comment (Kononnable) [inline docs/architecture/query_service.md:11]:
```suggestion
Communication to query service is done through [gRPC][grpc] based on two [endpoints][proto] of querying for data by `SCHEMA_ID` or multiple `OBJECT_ID`s. Query service communicates with multiple databases such as postgresql, druid, victoria metrics. Query service also communicates with [schema registry][schema-registry]. 
```
Sled is no longer there

File: docs/architecture/storage.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Storage Layer
+Consists of repositories for storing data.
+
+We can distinguish 3 types of supported repositories:
+- Document
+    - PostgreSQL
+- Blob
+    - Internal solution using [Sled][sled] database
+- Timeseries
+    - Victoria Metrics
```
Comment (Kononnable) [inline docs/architecture/storage.md:10]:
```suggestion
    - Druid
    - Victoria Metrics
```

File: docs/architecture/storage.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# Storage Layer
+Consists of repositories for storing data.
+
+We can distinguish 3 types of supported repositories:
+- Document
+    - PostgreSQL
+- Blob
+    - Internal solution using [Sled][sled] database
```
Comment (Kononnable) [inline docs/architecture/storage.md:8]:
```suggestion
```
currently archived

File: docs/architecture/storage.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+- Timeseries
+    - Victoria Metrics
+
+[sled]: https://github.com/spacejam/sled
```
Comment (Kononnable) [inline docs/architecture/storage.md:12]:
```suggestion
```

File: docs/architecture/storage.md [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# Storage Layer
+Consists of repositories for storing data.
+
+We can distinguish 3 types of supported repositories:
```
Comment (Kononnable) [inline docs/architecture/storage.md:4]:
```suggestion
Currently we support 2 types of repositories:
```

File: docs/deployment/docker-compose.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+There is also managament panel available at `localhost:15672`. The credentials are `user`/`CHANGEME`.
+By default there is no replication on *schema_registry*. Postgres *command_service* input channel is `cdl.document.data`.
+
+Errors are written to `cdl.reports` fanout exchange and can be read via `cdl.reports` queue.
+
+### PostgreSQL
+
+To access postgres you must have some postgresql client installed.
+
+For command line it's best to refer to your OS package manager (`homebrew` on OSX, `apt-get` on Ubuntu, `choco` on Windows).
```
Comment (Kononnable) [inline docs/deployment/docker-compose.md:65]:
```suggestion
For command line it's best to refer to your OS package manager (`homebrew` on OSX, `apt` on Ubuntu, `choco` on Windows).
```

File: docs/deployment/helm.md [hunk 10]
```diff
@@ -1,4 +1,4 @@
-# K8s local deployment
+# Helm
```
Comment (Kononnable) [inline docs/deployment/helm.md:1]: I'm not sure if this should be changed. This document describes local testing on k8s, not production deployment. We don't have a document on how to deploy cdl for production env.

File: docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg [hunk 11]
```diff
@@ -0,0 +1 @@
...service, data_type</text><polygon fill="#A80036" points="620,200.8281,630,204.8281,620,208.8281,624,204.8281" style="stroke: #A80036; stroke-width: 1.0;"/><line style="stroke: #A80036; stroke-width: 1.0;" x1="197.5" x2="626" y1="204.8281" y2="204.8281"/><text fill="#000000" font-family="sans-serif" font-size="13" lengthAdjust="spacingAndGlyphs" textLength="92" x="204.5" y="199.7622">Query for data</text><polygon fill="#A80036" points="208.5,229.9609,198.5,233.9609,208.5,237.9609,204.5,233.9609" style="stroke: #A80036; stroke-width: 1.0;"/><line style="stroke: #A80036; stroke-width: 1.0;" x1="202.5" x2="631" y1="233.9609" y2="233.9609"/><text fill="#000000" font-family="sans-serif" font-size="13" lengthAdjust="spacingAndGlyphs" textLength="121" x="214.5" y="228.895">Retrieve document</text><polygon fill="#A80036" points="63.5,259.0938,53.5,263.0938,63.5,267.0938,59.5,263.0938" style="stroke: #A80036; stroke-width: 1.0;"/><line style="stroke: #A80036; stroke-width: 1.0;" x1="57.5" x2="196.5" y1="263.0938" y2="263.0938"/><text fill="#000000" font-family="sans-serif" font-size="13" lengthAdjust="spacingAndGlyphs" textLength="121" x="69.5" y="258.0278">Retrieve document</text></g></svg>
```
Comment (Kononnable) [inline docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg:1]: Those files are generated automatically?

File: docs/schemas_and_views.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    "name": "two dimensions",
+    "definition": "{ x: x, y: y }"
+}
+```
+
+[indradb]: https://github.com/indradb/indradb
+[sled]: https://sled.rs/
+[jmespath]: https://jmespath.org/
+[json schema]: https://json-schema.org/
+[cdl cli]: architecture/cli.md
```
Comment (Kononnable) [inline docs/schemas_and_views.md:72]:
```suggestion
[jmespath]: https://jmespath.org/
[json schema]: https://json-schema.org/
```

**Review by Kononnable — COMMENTED** [review]
In general(not related to this PR) I would move configuration sections(environmental variable list) to a separate page.

File: docs/architecture/README.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+[frontend]   | Admin Web Panel - provides GUI interface for managing schemas and storing and retrieving data
+
+## GraphQL API
+[API] - used as a backend service for frontend, provides unified interface to manage CDL.
+
+## Configuration Layer
+Crate Name              | Purpose
+------------------------|--------
+[schema-registry]         | Manage user-defined schemas that define the format of incoming values and their respective topics
+[leader-elector]          | Elect master nodes in replicated services (_only for the Schema Repository, currently_)
```
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/README.md:22]: Imho architecture perspective is also important for contributing to the repo, and then you need to know what `leader-elector` is

File: docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg (hunk 11 above)
Comment (frondeus) [reply docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg:1]: Yes

File: docs/architecture/README.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+------------------------|--------
+[data-router]             | Route incoming data from and through MQ for consumption by the specific Command Service
+
+## Storage Layer
+
+Crate Name              | Purpose
+------------------------|--------
+[query-service]           | Wrap each individual repository for retrieval of data
+[command-service]         | Intake data from a MQ and storage, in specific repository
+[db-shrinker-storage]     | A service to remove older data from storage
```
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/README.md:35]: I dont understand, what repository name?

File: docs/architecture/README.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+------------------------|--------
+rpc                     | A collection of GRPC proto files and automatically generated client/server code.
+utils                   | A collection of utilities used throughout the Common Data Layer
+
+## Useful directories
+
+Directory       | Purpose
+----------------|--------
+deploy/helm     | helm charts for remote deployment
+deploy/compose  | sample deployment guide for docker
```
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/README.md:53]: Both helm and compose seems to be only for "development only".

File: docs/architecture/README.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+benchmarking    | scripts and scaffolding data for benchmarking
+tests           | component tests
+examples        | examplary client of cdl
+docs            | cdl documentation
+
+[cdl-cli]: cli.md
+[frontend]: frontend.md
```
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/README.md:60]: Agree. If the book goes first then #257 should change it. Otherwise I'll change this PR.

File: docs/architecture/cli.md [hunk 17]
```diff
@@ -9,10 +9,26 @@ For this tool to work, please make sure that the Schema Registry's gRPC server i
...
+To add a view under a schema already defined in the registry, run
+`cdl schema views -s <schema_name> add -n <view_name> -v <JMESPath_view>`. Views can only be added
+to schemas if they do not already exist on the schema; the `update` command can be used the same way
+as `add` to update existing schema views.
+
+_Make sure that views are valid [JMESPath](https://jmespath.org/) expressions._
+
+To list all views of a schema alphabetically, run `cdl schema views -s <schema_name> names`.
+
+To get a specific view on a schema, run `cdl schema views -s <schema_name> get -n <view_name>`.
```
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/cli.md:31]: We already have this info in README.md. I just moved it to new directory

File: docs/architecture/data_router.md [hunk 18]
```diff
@@ -75,4 +75,4 @@ Mind that GRPC uses HTTP2 as its transport protocol (L4), so SCHEMA_REGISTRY_ADD
 
 See an example [configuration][configuration] of deployment of data router and other services.
 
-[configuration]: ../deployment/compose/README.md
+[configuration]: ../deployment/docker-compose.md
```
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/data_router.md:78]: Maybe we should point to `./deployment/README.md`?

File: docs/deployment/helm.md (hunk 10 above)
Comment (frondeus) [reply docs/deployment/helm.md:1]: Probably we should introduce new folder `deployment/local` with `helm.md` and `docker-compose.md` inside, and then create `deployment/production` with `helm.md` only.

File: docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg (hunk 11 above)
Comment (frondeus) [reply docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg:1]: Question is - should we add it to gitignore then?

File: docs/architecture/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (Kononnable) [reply docs/architecture/README.md:35]: I mean repository as a component form architecture perspective(repository name might not be the best expression). All of those crates in storage layer with db engine are sometimes called a repository(e.g. in context of configuration service).
I think it would be good to say that storage layer(with db) is a repository(name which can be mentioned somewhere else in the docs), so people will know it's the same thing.

File: docs/architecture/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (Kononnable) [reply docs/architecture/README.md:53]: What do you mean by development only? We're officially supporting deployments based on helm, not docker-compose.
Are helm charts published separately? Nope, we're not there yet. However if someone wants to, they can set up production cluster based on those charts. I just wonder if we should mark it that docker deployment isn't officially supported - we don't guarantee it to be up to date(mentioned elsewhere in the docs, not sure if still true) or it performance.

File: docs/architecture/cli.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (Kononnable) [reply docs/architecture/cli.md:31]: Ok, so let me rephrase it. Should we remove it until #247 gets accepted(or implemented)?
I'm just pointing an issue to be discussed, I'm fine if it stays when we agree on that(that it should stay while functionality itself is WIP and this part will probably change).

File: docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg (hunk 11 above)
Comment (Kononnable) [reply docs/mdbook-plantuml-img/9208ca32e9197d5d57535abc78a61451de003302.svg:1]: I think we can - it's similar to js output from web-admin. Name is generated so I doubt anyone will look at it directly.

File: docs/README.md [hunk 19]
```diff
@@ -0,0 +1,5 @@
+# Common Data Layer
+
+The Common Data Layer (CDL) is a data storage service. It is designed with performance, versatility,
+scalability, and ease-of-modification as key tenets of its design, among others.
```
[the remark this answers is not in the document]
Comment (smores56) [reply docs/README.md:4]: In the future, please use this account (`smores56`) if you want me to get a notification.

I agree, generally it is good practice to avoid repetition with such a high degree of locality unless trying to make a point. In this case, the original sentence derives from the original documentation in Confluence written by Jakub. We can remove the "key tenets of its design" section, rendering something like:

```suggestion
The Common Data Layer (CDL) is a data storage service. Its primary goals are performance, versatility, scalability, and ease-of-modification.
```

**Comment by smores56** [comment]
Honestly, I have some clean up to do with the English, and believe it will take significantly less time to simply make a second PR once this gets merged and address it there, so I have avoided making comments on the English. I really appreciate the work to set up the structure of our documentation thus far, picking up my slack.

File: docs/architecture/README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (smores56) [reply docs/architecture/README.md:60]: Welp

**Comment by smores56** [comment]
Considering that the majority of this PR is the movement of the README's of services to this documentation, I suggest we remove the README's to avoid redundancy and to prevent them being out-of-sync with this documentation.

**Comment by frondeus** [comment]
> Considering that the majority of this PR is the movement of the README's of services to this documentation, I suggest we remove the README's to avoid redundancy and to prevent them being out-of-sync with this documentation.

I'd suggest keeping README but only with the link to the book chapter describing module.

File: docs/architecture/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/README.md:53]: I kept in mind:
> I'm not sure if this should be changed. This document describes local testing on k8s, not production deployment. We don't have a document on how to deploy cdl for production env.
it seems that both documents, `helm.md` and `docker-compose.md` describe non-production deployments.

[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/README.md:53]: Oh, okay now I get it, yes we should mark it as development only.

File: docs/architecture/cli.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (frondeus) [reply docs/architecture/cli.md:31]: I dont mind having a bit outdated docs and replace it after #247 gets accepted.

**Review by Kononnable — APPROVED** [review]
